Aggregate intervals from free and paid campaigns, render programs report w/o waiting for secondary request

tomalec opened the pull request:
### Changes proposed in this Pull Request:

- Aggregate intervals from free and paid campaigns. (034b9f1)
	Fixes #589
- Refactor `useProgramsReport` hook, to reduce unnecessary computation. (681e6bd)
	Memoize aggregated totals and intervals, re-calculate only when actually needed.
	
	Separate `secondary` report calls, so `Chart`, `CompareTable` could render earlier.
	Render `SummarySection` with only the `primary` data while `secondary` is still loading.
	Implements part of #654
- Remove unused mocked programs data (f8ebeb8)


### Screenshots:

paid and free program intervals are aggregated to render chart:
![Screencast showing how paid and free program intervals are aggregated to render chart correctly](https://user-images.githubusercontent.com/17435/119563562-fdca1700-bda7-11eb-8e76-56df1f04883e.gif)
UI is being rendered before secondary response arrives:
![Screencast showing how UI is being rendered before secondary response arrives](https://user-images.githubusercontent.com/17435/119571652-e4c66380-bdb1-11eb-9aef-96eccd53669f.gif)

### Detailed test instructions:

0. Make sure you have some test data, or mock it with 
	  ```php
	  add_filter(
	      'gla_prepared_response_ads-reports-products',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/products.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_mc-reports-products',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/products.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_ads-reports-programs',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_mc-reports-programs',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
	      },
	  );
	  ```
1. Open Programs report page [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms)
2. Switch between paid, free, and all programs and see that chart intervals add up.

---
Hard to observe without advanced mocking, as may be racy.
1. If the request for the previous period loads significantly longer than for the current one
2. You should see the entire root page be rendered, except deltas
3. Then update deltas eventually


### Things to consider:
- [ ] Add some tests for border cases and error cases, which are hard to catch using regular clicking without real data.
- [ ] Extend the enhancement to lazy load deltas to Products https://github.com/woocommerce/google-listings-and-ads/issues/654
- [ ] Add more explicit preloading of instead of `N/A` I'd also leave it to #654
- [ ] I hope this PR makes as closer not further away from https://github.com/woocommerce/google-listings-and-ads/issues/652

### Changelog Note:

> Fix programs report chart, aggregate intervals for free+paid programs.
> Improve UX by rendering programs report just after largest data chunk comes, without waiting for detlas.

tomalec committed:
Remove unused mocked programs data

tomalec committed:
Aggregate intervals from free and paid campaigns.

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/589

tomalec committed:
Refactor `useProgramsReport` hook, to reduce unnecesary computation.

Memoize aggregated totals and intervals, re-calculate only when actually needed.
For
https://github.com/woocommerce/google-listings-and-ads/issues/589

Separate `secondary` report calls, so `Chart`, `CompareTable` could render earlier.
Render `SummarySection` with only the `primary` data while `secondary` is still loading.
Implements part of https://github.com/woocommerce/google-listings-and-ads/issues/654

tomalec committed:
Fix blank page of doom, if secondary report comes with error,

or lacking some metrics.

eason9487 reviewed (commented):
📜 
It would not produce the duplicate API request if the consumer-side calls to `wp-data` selectors with the deep equaled parameters multiple times. So I think the `totals` (performance) data could be separated completely. I drafted a [diff](https://github.com/woocommerce/google-listings-and-ads/compare/37bf224a9bae1285e601de20575973f06943e653...ac4748a44b12b9b613f8eeb6b9baa99886dc6f0f) based on your branch to demonstrate that. Maybe it could be an alternative option for this PR.

And consider that the `totals` data would be the same when only changing the sorting-related queries. It should be easier to continue further improvements with the completely separate method when reducing API requests in the future.

tomalec committed:
Remove reduntand check in usePerformanceReport, fix default value of previousTotals

Address
https://github.com/woocommerce/google-listings-and-ads/pull/667#discussion_r639461327
https://github.com/woocommerce/google-listings-and-ads/pull/667#discussion_r639465470

tomalec committed:
Replace `reduce` with shorter `map` in `utils/aggregateIntervals`

Address https://github.com/woocommerce/google-listings-and-ads/pull/667#discussion_r639488584

tomalec committed:
Make `useProgramsReport` return always an array of `intervals`

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/667#discussion_r639434319

tomalec committed:
Fix `ProgramsReportSchema` JSDoc type definition.

tomalec committed:
Merge branch 'trunk' into fix/589-654-sort-intervals-when-needed

tomalec committed:
Fix `Array#map` syntax used in `aggregateIntervals`,

leftover after 7dcfb69460da94a74441a9f7d92d0e3dcffac0f9

eason9487 reviewed (approved):
Testing well and LGTM.
